pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+					Chain: externalTrafficChain,
+					Rule: knftables.Concat(
+						"goto", externalPolicyChain,
+					),
+				})
+			}
+		}
+
+		if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
+			// Generate the per-endpoint affinity sets
+			for _, ep := range allLocallyReachableEndpoints {
+				epInfo, ok := ep.(*endpointInfo)
+				if !ok {
+					klog.ErrorS(nil, "Failed to cast endpointsInfo", "endpointsInfo", ep)
+					continue
+				}
+
+				// Create a set to store current affinity mappings. The
+				// nft docs say "dynamic" is only needed for sets
+				// containing stateful objects (eg counters), but (at
+				// least on RHEL8) if we create the set without "dynamic",
+				// it later gets mutated to have it, and then the next
+				// attempt to tx.Add() it here fails because it looks like
+				// we're trying to change the flags.
+				tx.Add(&knftables.Set{
+					Name: epInfo.affinitySetName,
+					Type: ipvX_addr,
+					Flags: []knftables.SetFlag{
+						knftables.DynamicFlag,
+						knftables.TimeoutFlag,
+					},
+					Timeout: ptr.To(time.Duration(svcInfo.StickyMaxAgeSeconds()) * time.Second),
+				})
+				activeAffinitySets.Insert(epInfo.affinitySetName)
+			}
+		}
+
+		// If Cluster policy is in use, create the chain and create rules jumping
+		// from clusterPolicyChain to the clusterEndpoints
+		if usesClusterPolicyChain {
+			proxier.writeServiceToEndpointRules(tx, svcPortNameString, svcInfo, clusterPolicyChain, clusterEndpoints)
+		}
+
+		// If Local policy is in use, create rules jumping from localPolicyChain
+		// to the localEndpoints
+		if usesLocalPolicyChain {
+			proxier.writeServiceToEndpointRules(tx, svcPortNameString, svcInfo, localPolicyChain, localEndpoints)
+		}
+
+		// Generate the per-endpoint chains
+		for _, ep := range allLocallyReachableEndpoints {
+			epInfo, ok := ep.(*endpointInfo)
+			if !ok {
+				klog.ErrorS(nil, "Failed to cast endpointInfo", "endpointInfo", ep)
+				continue
+			}
+
+			endpointChain := epInfo.chainName
+
+			// Handle traffic that loops back to the originator with SNAT.
+			tx.Add(&knftables.Rule{
+				Chain: endpointChain,
+				Rule: knftables.Concat(
+					ipX, "saddr", epInfo.IP(),
+					"jump", kubeMarkMasqChain,
+				),
+			})
+
+			// Handle session affinity
+			if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
+				tx.Add(&knftables.Rule{
+					Chain: endpointChain,
+					Rule: knftables.Concat(
+						"update", "@", epInfo.affinitySetName,
+						"{", ipX, "saddr", "}",
+					),
+				})
+			}
+
+			// DNAT to final destination.
+			tx.Add(&knftables.Rule{
+				Chain: endpointChain,
+				Rule: knftables.Concat(
+					"meta l4proto", protocol,
+					"dnat to", epInfo.Endpoint,
+				),
+			})
+		}
+	}
+
+	// Figure out which chains are now stale. Unfortunately, we can't delete them
+	// right away because if there are map entries pointing to a chain, the chain
+	// doesn't become deletable until a short amount of time *after* the map entry is
```

Comment:
is this an async operation? is there more info on the nature of this "short amount of time"?

Reply:
Oh, I should have mentioned that this was a kernel bug and it's fixed in (very) recent kernels.

I don't have any more details (other than confirmation that, yes, it apparently did work the way I describe, and no, it wasn't intentional). If you look through the older e2e results and corresponding source pushes, you'll see it took me a few tries to fix this. Originally I thought you could do the deletes in a second transaction immediately after the first (ie as part of the same `syncProxyRules`), but that flakes. Then I tried doing them at the start of the _next_ `syncProxyRules`, but that flakes too because we configure the `BoundedFrequencyRunner` to allow "bursts", so you can get two syncs in a row before the rate-limiting kicks in. (I thought about changing that, except that I'm not sure if there's a really good reason why we do it that way, and anyway, an admin might configure a lower `minSyncPeriod` anyway...). So then that led to the current version, where we keep timestamps and only try to delete them after they've been stale for a whole second (which is probably orders of magnitude more time than we actually need to wait).

If we dug into the details of the kernel bug it might turn out that there would be a simpler way to handle the cleanup...

Reply:
Also, note that the timeout turns out to be too short in some circumstances, the result is just that you get warnings in the log and it tries to delete them again later. The stale chains don't actually hurt anything, other than taking up a tiny amount of memory and possibly confusing admins.